### PR TITLE
fix: binaryExpr not supported for LargeUtf8

### LINKED ISSUE #5893: like and ilike not supported for `LargeUtf8`
### Describe the bug

I'm trying to run an SQL query using ilike/like but it gives me the following error:
```
Data type LargeUtf8 not supported for scalar operation 'like' on string array. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```
The error seems to originate [here](https://github.com/apache/arrow-datafusion/blob/5c0fe0d7ebe0b205ce10c2b07ee2230ff3df6cba/datafusion/physical-expr/src/expressions/like.rs#L210)

### To Reproduce

given a table `test` with a column `a` of type `LargeUtf8` the following SQL query will fail
```sql
SELECT a FROM test WHERE "a" like '%something%';

SELECT a FROM test WHERE "a" ilike '%something%';
```

### Expected behavior

I'd expect to be able to do an ilike query on a `LargeUtf8` column.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #5893.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -233,7 +233,9 @@ macro_rules! compute_utf8_op_scalar {
             .as_any()
             .downcast_ref::<$DT>()
             .expect("compute_op failed to downcast left side array");
-        if let ScalarValue::Utf8(Some(string_value)) = $RIGHT {
+        if let ScalarValue::Utf8(Some(string_value))
+        | ScalarValue::LargeUtf8(Some(string_value)) = $RIGHT
+        {
```
Comment (jackwener) [inline datafusion/physical-expr/src/expressions/binary.rs:238]: I'm not sure if we should add it.

File: datafusion/core/tests/sqllogictests/test_files/select.slt [hunk 2]
```diff
@@ -230,3 +230,10 @@ select CASE
 END;
 ----
 2
+
+# like and ilike for LargeUtf8
+# issue: https://github.com/apache/arrow-datafusion/issues/5893
+query B
+select arrow_cast('foo', 'LargeUtf8') like '%foo%';
```
Comment (comphead) [inline datafusion/core/tests/sqllogictests/test_files/select.slt:237]: cool, btw, why is it like '%foo%', not == 'foo' ?

Comment (jackwener) [reply datafusion/core/tests/sqllogictests/test_files/select.slt:237]: It's a case of issue

Comment (comphead) [reply datafusion/core/tests/sqllogictests/test_files/select.slt:237]: > It's a case of issue

Sorry, I still didn't get it, you mean inner value of Utf8("foo") not the same as inner value LargeUtf8("foo")?

Comment (jackwener) [reply datafusion/core/tests/sqllogictests/test_files/select.slt:237]: Sorry, it's not clear.
This test cover issue in #5893

Comment (SimonSchneider) [reply datafusion/core/tests/sqllogictests/test_files/select.slt:237]: I guess a more comprehensive test would be 
```
select arrow_cast('foobar', 'LargeUtf8') like '%oba%';
—-
true


select arrow_cast('FooBar', 'LargeUtf8') like '%oba%';
—-
true
```


https://www.postgresql.org/docs/7.3/functions-matching.html#:~:text=The%20keyword%20ILIKE%20can%20be,and%20~~*%20corresponds%20to%20ILIKE%20.

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary.rs:238]: I definitely think we should add this -- it is required to evaluate the like expression with a scalar.

File: datafusion/core/tests/sqllogictests/test_files/select.slt (hunk 2 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/select.slt:237]: Another test that would be great to add would be to compare columns as well as constants (as that is the other binary expr codepath)

Something like

```sql
create table t as values 
  (arrow_cast('Foo', 'LargeUtf8') '%f'), 
  (arrow_cast('Bar', 'LargeUtf8') 'B%');

select column1 like column2 from t;
select column1 ilike column2 from t;
select column1 not like column2 from t;
select column1 not ilike column2 from t;
```

**Review by alamb — COMMENTED** [review]
Thank you for working on this @jackwener

**Comment by mingmwang** [comment]
I remember there are some other issues with `LargeUtf8`, for example, can not `Sort`
Maybe we need another unit test suite to cover all different Exprs with `LargeUtf8`.

**Comment by mingmwang** [comment]
```rust

#[tokio::test]
async fn sort_on_window_null_string() -> Result<()> {
    let d1: DictionaryArray<Int32Type> =
        vec![Some("one"), None, Some("three")].into_iter().collect();
    let d2: StringArray = vec![Some("ONE"), None, Some("THREE")].into_iter().collect();
    let d3: LargeStringArray =
        vec![Some("One"), None, Some("Three")].into_iter().collect();

    let batch = RecordBatch::try_from_iter(vec![
        ("d1", Arc::new(d1) as ArrayRef),
        ("d2", Arc::new(d2) as ArrayRef),
        ("d3", Arc::new(d3) as ArrayRef),
    ])
    .unwrap();

    let ctx = SessionContext::with_config(SessionConfig::new().with_target_partitions(1));
    ctx.register_batch("test", batch)?;

    let sql =
        "SELECT d1, row_number() OVER (partition by d1) as rn1 FROM test order by d1 asc";

    let actual = execute_to_batches(&ctx, sql).await;
    // NULLS LAST
    let expected = vec![
        "+-------+-----+",
        "| d1    | rn1 |",
        "+-------+-----+",
        "| one   | 1   |",
        "| three | 1   |",
        "|       | 1   |",
        "+-------+-----+",
    ];
    assert_batches_eq!(expected, &actual);

    let sql =
        "SELECT d2, row_number() OVER (partition by d2) as rn1 FROM test ORDER BY d2 asc";
    let actual = execute_to_batches(&ctx, sql).await;
    // NULLS LAST
    let expected = vec![
        "+-------+-----+",
        "| d2    | rn1 |",
        "+-------+-----+",
        "| ONE   | 1   |",
        "| THREE | 1   |",
        "|       | 1   |",
        "+-------+-----+",
    ];
    assert_batches_eq!(expected, &actual);

    let sql =
        "SELECT d2, row_number() OVER (partition by d2 order by d2 desc) as rn1 FROM test ORDER BY d2 desc";

    let actual = execute_to_batches(&ctx, sql).await;
    // NULLS FIRST
    let expected = vec![
        "+-------+-----+",
        "| d2    | rn1 |",
        "+-------+-----+",
        "|       | 1   |",
        "| THREE | 1   |",
        "| ONE   | 1   |",
        "+-------+-----+",
    ];
    assert_batches_eq!(expected, &actual);

    // FIXME sort on LargeUtf8 String has bug.
    // let sql =
    //     "SELECT d3, row_number() OVER (partition by d3) as rn1 FROM test";
    // let actual = execute_to_batches(&ctx, sql).await;
    // let expected = vec![
    //     "+-------+-----+",
    //     "| d3    | rn1 |",
    //     "+-------+-----+",
    //     "|       | 1   |",
    //     "| One   | 1   |",
    //     "| Three | 1   |",
    //     "+-------+-----+",
    // ];
    // assert_batches_eq!(expected, &actual);

    Ok(())
}
```

**Comment by jackwener** [comment]
@mingmwang Can we add it in following PR？

**Comment by jackwener** [comment]
if there isn't new review, I prepare to merge this PR in the tomorrow due to some review already in this PR.

**Comment by alamb** [comment]
I agree adding additional tests for unrelated functionality (e.g. general support for LargeUtf8) would be best done in some other PR

**Pull request merged**